### PR TITLE
Refactor: Logging System

### LINKED ISSUE #5468: Better logging system
Currently, LMMS uses several functions to print logs, e.g., `printf()`, `fprintf()`, `qWarning()`, etc. It has several issues and can be improved.

- [ ] Print log into a file (#3876)
- [ ] Better and unified logging API (#5159)
- [ ] Log viewer in the GUI

Please add comments if you have any suggestions.

### PR DESCRIPTION
**Pull request by TechnoPorg**
Closes #5468 
Supersedes #5500

### Brief Overview
LMMS' logging system is currently a mix of printf, and QT debug messages. Additionally, logs aren't saved to a file, and can't be easily written to console on Windows, which makes it both challenging for the average user to put anything in the "Relevant Log Output" section of a bug report, and more frustrating for developers attempting to track down bugs and ensure consistent behaviour.

This PR implements a new central logging system that is intended to be flexible in its capabilities and usable in all parts of the application, including realtime worker threads.

### Slightly adapted description from the original PR:

> The new API is intended to be used universally in the application, including the performance-critical sections.
> 
> Basics:
> 
>  - The central class is LogManager, which needs to be accessible from any place in the code. To achieve this, it was implemented as a singleton.
>  - The LogManager stores a list of LogSinks. Each LogSink represents a different kind of output for the log lines. Examples are: standard output, a log file, a debug messages' window inside an application, a remote endpoint...
>  - Each LogSink is assigned independently the highest level of verbosity of the messages. Currently there are five levels: Fatal, Error, Warning, Info, Trace.
>  - Whenever the LogManager accepts a log line, it is broadcast to all the sinks, unless the verbosity level of the line is greater than the configured verbosity level. For example, if we set the log verbosity level to Warning, all Errors will appear in both places. On the other hand, Infos will not be present.
>  - In order to shorten the code, the programmer is supposed to use a set of dedicated macros rather than calling LogManager's API directly. There are two sets of macros. LOG_xxx macros accept the same syntax as printf function, while LOG_XXX_TOPIC macros are supposed to be used as C++ streams (or Qt's mechanism). Replace "xxx" by the verbosity level designator: FATAL, ERR, WARN, INFO, TRACE.
>  - Each log line, besides the actual message, is accompanied by some metadata: an optional topic, verbosity information, a timestamp and the location of the log (source file name and line number). The timestamps have a resolution of 1 ms.

### Things I did differently (potentially questionable choices):
- Put the logging files under `/log/` directories. I was trying not to be a part of the problem of LMMS having a single monolithic include folder, in the spirit of #5944, but I realize that it's quite a significant break in convention, so if you're not on board with it, I'll move everything to be more in line with the existing structure. `LoggingMacros.h`, the only part that almost all logging functionality will use, is under the main folder to make it "more accessible".
- Removed the ability to pause and disable the logging thread. Given that buffers will always be flushed on a fatal log message or on shutdown, and the logging is done on a low-priority thread, I really don't think this has any benefit for either resilience or performance.
- Removed the per-sink max verbosity level, because I just don't see the practical use. I think if someone really wants this in a custom sink, they can implement it themselves. 
- Trace messages do not show up in release builds for performance reasons.

### Some actionable items copied from the original PR:
- implementation of sinks for log file ~~and log window~~
- adding appropriate configuration options, accessible via GUI _(Log output file location and max log verbosity)_
- enhancing log messages with the information about the name of a thread that generated each line _(Although I don't believe this to be necessary)_
- usage of smart pointers instead of C-style pointers to store the sinks

### And some things I'd like to see considered:
- Remove the LogTopic system. @JohannesLorenz I know you were in favour of it for the initial PR, but I feel like in practice, it adds a lot of complication to the code, and developers are just going to grep logs by filename anyways.
- Add far more extensive log calls to existing parts of the code, and have more requirements that new code include descriptive logging.

### CODE REVIEW

**Comment by TechnoPorg** [comment]
... except as I was climbing into bed I realized it's not realtime safe because log lines are allocated, plus a buffer for formatted strings. Let me sleep on this one.

**Comment by TechnoPorg** [comment]
I've fixed the build errors (printf being called without string literals and MinGW threads), and made it so that logging is realtime safe in terms of string formatting. Note that if a log message has format arguments, it will no longer be realtime safe.

I'm still not sure what to do about the memory allocation of log lines. I'm not sure how this made it through the original PR unnoticed, honestly, but I can't find a way around it with the ringbuffer implementation. Any thoughts?

**Comment by Snowiiii** [comment]
I think we should use the Qt Logging System, And it is very easy to implement File logging in there, Here in an full guide: https://doc.qt.io/qt-6/qtlogging.html

**Comment by TechnoPorg** [comment]
> I think we should use the Qt Logging System, And it is very easy to implement File logging in there, Here in an full guide: [doc.qt.io/qt-6/qtlogging.html](https://doc.qt.io/qt-6/qtlogging.html)

The Qt logging system is very powerful, and it's a good idea, but there are two reasons why it would not be a good fit for most of LMMS' logging needs (See the original PR for a detailed discussion of these):
1. It is not realtime safe. The goal in implementing a custom logging system for LMMS is to be able to log from audio worker threads without stutters, which this PR will be able to accomplish, but Qt cannot do.
2. We want to get Qt out of the application core, and using the Qt logging system would bring heavier coupling with it. A solution using Qt logging would eventually have to be rewritten to not use Qt anyways.

**Comment by sakertooth** [comment]
I can't help but think this implementation drives up so much complexity, compared to simply designing the code in such a way that the audio threads don't have to do any logging at all. When I took a look at the code, I had to jump through many different files to really understand that it was all about using a lock-free ring buffer to do the communication.

And this will not always be running in real-time. The use of `new` and `delete` for creating a `LogLine` is already a problem, not to mention the conditional flushes that may happen in `LogManager::flush`, which would eventually lead to the audio thread interacting with the I/O using `std::cout`, i.e., what we wanted to avoid in the first place.

I also have a feeling that the simple use of `std::cout` would be more performant than this (not necessarily in terms of it being in real-time, but in general), just given that the interactions are probably more simpler, but I cant say for sure because there aren't any measurements.

The design's complexity is a bit much too. `ConsoleLogSink` is the only thing implementing `LogSink`, which feels a bit redundant. This decision seems like it was based on a effort to make things more decoupled, but in doing so it drives up complexity.

In summary, I would design the code in such a way that the audio threads that are meant to be running in real-time never have to worry about logging in the first place. If there is no choice, then I wouldn't see the problem with printing log messages only when using a Debug build, meaning that our Release builds wont actually be doing any logging to begin with, making the Release build (i.e., what will most likely be used by people actually using the program), more real-time safe as far as logging is concerned.

**Comment by Veratil** [comment]
Don't want to sound like I'm dismissing the work done, but instead of writing our own I think we should try to use a dedicated logging library we can drop in. For example something like [xtr](https://github.com/choll/xtr) (found by searching `c++ logging realtime`).

**Comment by TechnoPorg** [comment]
> And this will not always be running in real-time. The use of new and delete for creating a LogLine is already a problem, not to mention the conditional flushes that may happen in LogManager::flush, which would eventually lead to the audio thread interacting with the I/O using std::cout, i.e., what we wanted to avoid in the first place.

Yeah, I was looking for input about how those new/delete calls could be removed.

> In summary, I would design the code in such a way that the audio threads that are meant to be running in real-time never have to worry about logging in the first place.

Fair enough. The only people who are really going to care about what the audio thread is doing already know enough to use GDB.

> Don't want to sound like I'm dismissing the work done, but instead of writing our own I think we should try to use a dedicated logging library we can drop in. For example something like [xtr](https://github.com/choll/xtr) (found by searching c++ logging realtime).

I didn't take it that way, don't worry! I really should have thought this through and discussed it before sinking my time into something that's not really the right fit for the software... I wasn't aware of the existence of xtr, but I agree that it is a better fit than whatever this chimera of a PR is.

----------------------
So overall, I still fully believe LMMS needs a centralized logging system as laid out in #5468, but having listened to what you two have to say about this one, I realise that it's the wrong way to go about it. I think the ideas of verbosity levels and file/timestamp metadata are worth keeping, but the rest I am willing to throw out in favour of a 3rd party library like XTR, Log4CXX, etc.

Would you like me to close this PR?

**Comment by Snowiiii** [comment]
Would recommend spdlog its one of the popularist logging libraries and has great performance, It has build in File Logger also

**Comment by sakertooth** [comment]
> Don't want to sound like I'm dismissing the work done, but instead of writing our own I think we should try to use a dedicated logging library we can drop in. For example something like [xtr](https://github.com/choll/xtr) (found by searching c++ logging realtime).

One small problem though... XTR doesn't support Windows.

From their README:
> #### Supported platforms
>  * Linux (x86-64)
>  * FreeBSD (x86-64)

Otherwise I would've definitely said go for it.

> Would recommend spdlog its one of the biggest logging libraries and has great performance, It has build in File Logger also

spdlog didn't make any guarantees on if it can be used in real-time. I'm leaning more towards XTR because of this for our audio threads, but it doesn't have support for Windows.

**Comment by Snowiiii** [comment]
Can you may dig abit into spdlog, Its just an Trusted good library which i already used myself, Im always a bit cautious with these smaller libs like XTR

**Comment by Veratil** [comment]
I completely missed the no Windows support for xtr. Maybe it could be something we could look into adding support for?

I saw spdlog, but looking at the benchmarks in xtr's table it's very slow. [quill](https://github.com/odygrd/quill) has Windows support.

**Comment by sakertooth** [comment]
Quill might be it then.

**Comment by TechnoPorg** [comment]
Closing this as it has been determined to not be the right solution.

If someone could also close #5500 that would be great!
